### PR TITLE
[SYCL-MLIR]: Support option -w (suppress all warnings)

### PR DESCRIPTION
**Pull request by etiotto**
This PR adds support for the -w options to cgeist. This option can be used to suppress compiler warnings. The clang driver is modified to pass the -w option to cgeist.

### CODE REVIEW

**Comment by etiotto** [comment]
Note: currently `cgeist` emits a number of warning messages to report missing functionality.  I have added support for the -w option to filter out those msgs (although the -w is a general useful option).

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -436,8 +437,10 @@ static const clang::ConstantArrayType *getCAT(const clang::Type *T) {
 
 ValueCategory MLIRScanner::VisitArrayInitLoop(clang::ArrayInitLoopExpr *expr,
                                               ValueCategory tostore) {
+  if (!SuppressWarnings)
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:440]: Do you think it make sense to have a function like:
```
void warn(const Twine &msg) {
  if (SuppressWarnings)
    return;

  llvm::WithColor::warning() << msg;
}
```
and replace all these warnings to calling this function?

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -1170,12 +1165,9 @@ ValueCategory MLIRScanner::VisitAtomicExpr(clang::AtomicExpr *BO) {
...
 ValueCategory MLIRScanner::VisitExprWithCleanups(ExprWithCleanups *E) {
   auto ret = Visit(E->getSubExpr());
-  for (auto &child : E->children()) {
-    if (!SuppressWarnings) {
-      child->dump();
-      llvm::WithColor::warning() << "cleanup not handled\n";
-    }
-  }
+  for (auto &child : E->children())
+    mlirclang::warning() << "cleanup not handled for: " << child << "\n";
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1170]: `child->dump()` and `<< child` may actually print different results, can we do something like 
`child->dump(mlirclang::warning())`?

File: polygeist/tools/cgeist/Lib/utils.cc [hunk 3]
```diff
@@ -91,3 +94,9 @@ mlir::gpu::GPUModuleOp mlirclang::getDeviceModule(mlir::ModuleOp Module) {
   return cast<mlir::gpu::GPUModuleOp>(
       Module.lookupSymbol(MLIRASTConsumer::DeviceModuleName));
 }
+
+llvm::raw_ostream &mlirclang::warning() {
+  if (SuppressWarnings)
+    return llvm::nulls();
+  return llvm::WithColor::warning();
+}
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/utils.cc:102]:
```suggestion
}

```

File: polygeist/tools/cgeist/Lib/utils.cc [hunk 4]
```diff
@@ -11,25 +11,28 @@
...
 
 #include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/StringRef.h"
-
-#include "mlir/Dialect/Linalg/IR/Linalg.h"
-#include "mlir/Dialect/MemRef/IR/MemRef.h"
-
-#include "clang/AST/Expr.h"
+#include "llvm/Support/WithColor.h"
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/utils.cc:27]: We should be able to remove this header from some files now?

**Review by victor-eds — APPROVED** [review]
LGTM after Whitney's comments.

File: polygeist/tools/cgeist/Lib/utils.cc (hunk 4 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/utils.cc:27]: not from this one because it is used in the `warning()` function

Comment (etiotto) [reply polygeist/tools/cgeist/Lib/utils.cc:27]: I already removed the #include for that header in other places though.

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/utils.cc:27]: Yup, see them on the new commit, so resolved this conversation.

File: polygeist/tools/cgeist/Lib/utils.h [hunk 5]
```diff
@@ -46,10 +44,10 @@ namespace mlirclang {
 /// operands %a and %b. The new op will be inserted at where the insertion point
 /// of the provided OpBuilder is.
 mlir::Operation *
-replaceFuncByOperation(mlir::func::FuncOp f, llvm::StringRef opName,
-                       mlir::OpBuilder &b,
-                       llvm::SmallVectorImpl<mlir::Value> &input,
-                       llvm::SmallVectorImpl<mlir::Value> &output);
+replaceFuncByOperation(mlir::func::FuncOp Func, llvm::StringRef OpName,
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/utils.h:47]: Sorry. Is a moot point I have another PR (https://github.com/intel/llvm/pull/7492) to clean the code up. If that lands before this one all will be good.

[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/utils.h:47]: ok, let me review #7492 now.

**Review by whitneywhtsang — APPROVED** [review]
Please merge in https://github.com/intel/llvm/pull/7492 first.

**Pull request merged**